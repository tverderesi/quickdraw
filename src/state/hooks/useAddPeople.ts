import { useRecoilState, useSetRecoilState } from "recoil";
import { errorState, peopleState } from "../atoms";
import { InputError } from "../errors/inputError";

export const useAddPeople = () => {
  const [people, setPeople] = useRecoilState(peopleState);
  const setError = useSetRecoilState(errorState);
  return (name: string) => {
    const isInList = people.includes(name);
    try {
      if (isInList) {
        throw new InputError("Duplicate name");
      } else {
        setPeople((people: string[]) => [...people, name]);
      }
    } catch (err: any) {
      setError(err);
      setTimeout(() => {
        setError(null);
      }, 5000);
      console.error(err);
    }
  };
};
