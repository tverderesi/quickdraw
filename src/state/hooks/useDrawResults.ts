import { resultState } from "../atoms";
import { useRecoilValue } from "recoil";

export const useDrawResults = () => {
  return useRecoilValue(resultState);
};
