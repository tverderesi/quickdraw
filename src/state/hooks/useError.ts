import { useRecoilValue } from "recoil";
import { errorState } from "../atoms";

export const useError = () => {
  return useRecoilValue(errorState);
};
