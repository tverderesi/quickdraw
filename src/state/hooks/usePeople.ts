import { useRecoilValue } from "recoil";
import { peopleState } from "../atoms";

export const usePeople = () => {
  return useRecoilValue(peopleState);
};
