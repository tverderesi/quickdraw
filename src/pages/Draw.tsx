import { usePeople } from "../state/hooks/usePeople";
import { ButtonHTMLAttributes, useState, MouseEventHandler } from "react";
import { useDrawResults } from "../state/hooks/useDrawResults";
import OouiArrowPreviousLtr from "../components/OouiArrowPreviousLtr";
import OouiArrowNextLtr from "../components/OouiArrowNextLtr";

export default function Draw() {
  const people = usePeople();

  const [secretSanta, setSecretSanta] = useState("");
  const [index, setIndex] = useState(0);

  const result = useDrawResults();

  const handleNavClick = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>,

    left: boolean
  ) => {
    left ? setIndex(index + 1) : setIndex(index - 1);
  };

  const handleClick = () => {
    result.has(people[index]) && setSecretSanta(result.get(people[index])!);
    setTimeout(() => {
      setSecretSanta("");
    }, 2000);
  };
  return (
    <section className="relative card py-6 px-3 bg-info shadow-2xl shadow-neutral m-5 lg:m-10 text-info-content h-[calc(100vh-17.5rem)] ">
      <h2 className="text-5xl text-center w-full">Next Person</h2>
      <button
        disabled={!index}
        onClick={(e) => handleNavClick(e, false)}
        className="btn btn-square btn-secondary text-3xl absolute left-5 transition-all duration-200 ease-in-out"
      >
        <OouiArrowPreviousLtr />
      </button>
      <button
        disabled={index >= people.length - 1}
        onClick={(e) => handleNavClick(e, true)}
        className="btn btn-square btn-secondary text-3xl absolute right-5 transition-all duration-200 ease-in-out"
      >
        <OouiArrowNextLtr />
      </button>
      <p className="text-7xl font-semibold w-full text-center py-10">
        {people[index]}
      </p>

      <button onClick={handleClick} className="btn btn-xl mx-auto">
        Draw
      </button>

      {secretSanta && (
        <div className="flex flex-row items-center justify-center">
          <div
            className="alert alert-warning shadow-lg w-96 m-5 absolute bottom-3 z-10 transition-all duration-300 ease-in-out"
            role="alert"
          >
            <div>
              <svg
                xmlns="http://www.w3.org/2000/svg"
                className="stroke-current flex-shrink-0 h-6 w-6"
                fill="none"
                viewBox="0 0 24 24"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
                />
              </svg>
              <span className="text-2xl w-full text-center">
                Your secret santa is{" "}
                <span className="font-bold">{secretSanta}</span>!
              </span>
            </div>
          </div>
        </div>
      )}
    </section>
  );
}
