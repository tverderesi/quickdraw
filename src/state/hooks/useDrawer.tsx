import { useSetRecoilState } from "recoil";
import { resultState } from "../atoms";
import { draw } from "../../util/draw";
import { usePeople } from "./usePeople";

export const useDrawer = () => {
  const participants = usePeople();
  const setResult = useSetRecoilState(resultState);
  return () => {
    const result = draw(participants);
    setResult(result);
  };
};
