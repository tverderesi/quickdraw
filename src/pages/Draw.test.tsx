import { fireEvent, render, screen } from "@testing-library/react";
import { RecoilRoot } from "recoil";
import { usePeople } from "../state/hooks/usePeople";
import { useDrawResults } from "../state/hooks/useDrawResults";
import Draw from "./Draw";

jest.mock("../state/hooks/usePeople", () => ({
  usePeople: jest.fn(),
}));

jest.mock("../state/hooks/useDrawResults", () => ({
  useDrawResults: jest.fn(),
}));

describe("on the draw page", () => {
  const participants = [
    "Anne",
    "Catherine",
    "Jackson",
    "July",
    "Joseph",
    "Nathalia",
  ];
  const results = new Map([
    ["Anne", "Joseph"],
    ["Catherine", "July"],
    ["Jackson", "Nathalia"],
    ["July", "Anne"],
    ["Joseph", "Catherine"],
    ["Nathalia", "Jackson"],
  ]);

  beforeEach(() => {
    (usePeople as jest.Mock).mockReturnValue(participants);
    (useDrawResults as jest.Mock).mockReturnValue(results);
  });

  test("the Secret Santa is displayed when requested", () => {
    render(
      <RecoilRoot>
        <Draw />
      </RecoilRoot>
    );

    const button = screen.getByRole("button", { name: "Draw" });
    fireEvent.click(button);

    const secretSanta = screen.getByRole("alert");
    expect(secretSanta).toMatchSnapshot();
  });
});
