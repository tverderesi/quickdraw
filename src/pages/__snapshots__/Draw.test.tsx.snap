// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on the draw page the Secret Santa is displayed when requested 1`] = `
<div
  class="alert alert-warning shadow-lg w-96 m-5 absolute bottom-3 z-10 transition-all duration-300 ease-in-out"
  role="alert"
>
  <div>
    <svg
      class="stroke-current flex-shrink-0 h-6 w-6"
      fill="none"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    <span
      class="text-2xl w-full text-center"
    >
      Your secret santa is
       
      <span
        class="font-bold"
      >
        Joseph
      </span>
      !
    </span>
  </div>
</div>
`;
